go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -69,14 +68,127 @@ type NetworkControllerManager struct {
+	p1 := func(item *nbdb.LogicalSwitch) bool {
+		_, ok := item.ExternalIDs[ovntypes.NetworkNameExternalID]
+		return ok
+	}
+	nodeSwitches, err := libovsdbops.FindLogicalSwitchesWithPredicate(nbClient, p1)
+	if err != nil {
+		klog.Errorf("Failed to get all logical switches of secondary network error: %v", err)
+		return nil, nil, err
+	}
+	p2 := func(item *nbdb.LogicalRouter) bool {
+		_, ok := item.ExternalIDs[ovntypes.NetworkNameExternalID]
+		return ok
+	}
+	clusterRouters, err := libovsdbops.FindLogicalRoutersWithPredicate(nbClient, p2)
+	if err != nil {
+		klog.Errorf("Failed to get all distributed logical routers: %v", err)
+		return nil, nil, err
+	}
+	return nodeSwitches, clusterRouters, nil
+}
+
+func (cm *networkControllerManager) SyncNetworkControllers(allControllers []NetworkController) error {
+	existingNetworksMap := map[string]bool{}
+	for _, oc := range allControllers {
+		existingNetworksMap[oc.GetNetworkName()] = true
+	}
+
+	// Get all the existing secondary networks and its logical entities
+	switches, routers, err := findAllSecondaryNetworkLogicalEntities(cm.nbClient)
+	if err != nil {
+		return err
+	}
+
+	var ops []ovsdb.Operation
+	staleNetworks := map[string]NetworkController{}
+	for _, ls := range switches {
+		netName := ls.ExternalIDs[ovntypes.NetworkNameExternalID]
+		if _, ok := existingNetworksMap[netName]; ok {
+			// network still exists, no cleanup to do
+			continue
+		}
+		if topoType, ok := ls.ExternalIDs[ovntypes.TopoTypeExternalID]; ok {
+			// Create dummy network controllers to clean up logical entities
+			klog.V(5).Infof("Found stale %s network %s", topoType, netName)
+			if oc, err := cm.NewDummyNetworkController(topoType, netName); err == nil {
+				staleNetworks[netName] = oc
+				continue
+			}
+		}
+		klog.Infof("Missing %s external-id on switch %s, simply deleted it", ovntypes.TopoTypeExternalID, ls.Name)
+		ops, err = libovsdbops.DeleteLogicalSwitchOps(cm.nbClient, ops, ls.Name)
+		if err != nil {
+			klog.Errorf("Failed to get ops to delete stale logical switch %s for network %s: %v", ls.Name, netName, err)
+		}
+	}
+	for _, lr := range routers {
+		netName := lr.ExternalIDs[ovntypes.NetworkNameExternalID]
+		if _, ok := existingNetworksMap[netName]; ok {
+			// network still exists, no cleanup to do
+			continue
+		}
+		if topoType, ok := lr.ExternalIDs[ovntypes.TopoTypeExternalID]; ok {
+			// Create dummy network controllers to clean up logical entities
+			klog.V(5).Infof("Found stale %s network %s", topoType, netName)
+			if oc, err := cm.NewDummyNetworkController(topoType, netName); err == nil {
+				staleNetworks[netName] = oc
+				continue
+			}
+		}
+		klog.Infof("Missing %s external-id on router %s, simply deleted it", ovntypes.TopoTypeExternalID, lr.Name)
+		ops, err = libovsdbops.DeleteLogicalRouterOps(cm.nbClient, ops, lr)
+		if err != nil {
+			klog.Errorf("Failed to get ops to delete logical router %s for network %s: %v", lr.Name, netName, err)
+		}
```

Comment:
this is done only when we don't know how to create the dummyNetwokController to do later cleanup (because we don't know its topology type).  why cannot do it here?

Reply:
I see. Why would that happen? Isn't the `TopoTypeExternalID` being set in the same operation as the `NetworkNameExternalID`?

Reply:
this should not happen, but I just do not want to blindly assume that being set in the code.